### ISSUE TITLE
SMT encoding can produce invalid patterns

### ISSUE THREAD

**Issue by tchajed** [FStarLang/FStar#1197]
The following example results in `./invalid_function_app.fst(0,0-0,0): (Warning) Invalid_function_app: Unexpected output from Z3: (error "line 43467 column 13: invalid function application, arguments missing")`. Z3 encounters an error (which is actually in a pattern), which becomes an F* warning due to unexpected output from Z3.

```
module Invalid_function_app

assume val p : Type0

assume val prove_p : #t:Type -> a:t -> Lemma p
    [SMTPat (a)]

let use_prove_p (u:unit) : Lemma p = prove_p ()
```

**Comment by aseemr** [comment]
Fixed in 19fbdba4b365683afb84d39c776c1ababdd746d1.

The compiler should have given a warning, since the pattern does not contain all bound variables (`t`). The commit above fixes it, essentially, for `Prims.smtpat #a b`, it does not consider `#a` when computing the free variables of the pattern, since `#a` is not encoded to the solver.

**Comment by nikswamy** [comment]
Thanks for @R1kM 

```module LowStar.RST.Par

open LowStar.Resource
open LowStar.RST

assume
val par (#in1 #in2:resource)
        (#a #b:Type)
        (#out1:a -> resource)
        (#out2:b -> resource)
        (#pre1:r_pre in1)
        (#pre2:r_pre in2)
        (#post1:r_post in1 a out1)
        (#post2:r_post in2 b out2)
        ($f1:unit -> RST a in1 out1 pre1 post1)
        ($f2:unit -> RST b in2 out2 pre2 post2)
        : RST (a & b)
              (in1 <*> in2)
              (fun (x, y) -> out1 x <*> out2 y)
              (fun h -> pre1 h /\ pre2 h)
              (fun h0 (x, y) h1 -> post1 h0 x h1 /\ post2 h0 y h1)

let f () : RST unit (empty_resource) (fun _ -> empty_resource) (fun _ -> True) (fun _ _ _ -> True) = ()

let test (b1 b2:array UInt32.t) : RST unit
  (empty_resource <*> empty_resource)
  (fun _ -> empty_resource <*> empty_resource)
  (fun _ -> True)
  (fun _ _ _ -> True)
  = 
  // The call to par succeeds (we can verify with an admit just after), 
  // We can also add a pure let binding for instance just after, 
  // but chaining it with another stateful call or returning unit fails
  let x = par f f in
  ()
```

Produces `Unexpected output from Z3: WARNING: (284560, 57): pattern does not contain all quantified variables`

**Comment by R1kM** [comment]
The modules LowStar.Resource and LowStar.RST are available in afromher_lowstar_lens. Permanant links below:

https://github.com/FStarLang/FStar/blob/e2163a558c384602b30756abbb591738c85ea451/ulib/LowStar.RST.fst
https://github.com/FStarLang/FStar/blob/e2163a558c384602b30756abbb591738c85ea451/ulib/LowStar.Resource.fst

**Comment by R1kM** [comment]
After some more experimentation, the destruction of the pair in the clause
`             (fun (x, y) -> out1 x <*> out2 y)`
seems to be what is failing. If I replace it by
` (fun p -> out1 (fst p) <*> out2 (snd p))`, this error disappears

**Comment by aseemr** [comment]
The problematic pattern is coming up in one of the partial application typing axioms, which are gone from the F* master branch. So, @R1kM, if you merge master into your branch, you should no longer see this.

I think @nikswamy  fixed this issue in the #1750 branch that we were working on, before these axioms were nuked altogether.

Closing the issue, please reopen if you observe this even after merging with master. Thanks for the report!
